### ISSUE TITLE
[Bug]: Shaders Issues

### ISSUE THREAD

**Issue by Zbuffers** [AssetRipper/AssetRipper#1040]
### Are you on the latest version of AssetRipper?

Yes, I'm on the latest release of AssetRipper.

### Which release are you using?

Windows x64

### Which game did this occur on?

[Game](https://steamdb.info/app/880940/)

### Which Unity version did this occur on?

2021.3.30

### Is the game Mono or IL2Cpp?

Mono

### Describe the issue.

Hi,

I think these issues are 'partly' already known, but I think I faced a new one (See case N°3 YAML Shader error in Unity)
So I thought that I would report them all in one issue.
Feel free to change/close this issue if it's already reported.
Attached are 5 files (3 logs, the exported 2021.3.30 YAML shader, and the 2020.3.35 YAML shader that imports fine)

When trying to export a (single or all) shaders, I face various issues:

1. using Decompile mode to export 1 shader, I get a crash during export:
```
System.IO.EndOfStreamException: Can't read 1869367076-byte string because there are only 393832 bytes left in the stream
   at AssetRipper.IO.Endian.EndianReader.ReadString(Int32 length) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.IO.Endian\EndianReader.cs:line 217
   at AssetRipper.Assets.IO.Reading.AssetReader.ReadString() in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Assets\IO\Reading\AssetReader.cs:line 22
   at AssetRipper.IO.Endian.EndianReader.ReadStringArray(Boolean allowAlignment) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.IO.Endian\EndianReader.cs:line 588
   at AssetRipper.IO.Endian.EndianReader.ReadStringArray() in F:\_DEV2\AssetRipper-master\Source\AssetRipper.IO.Endian\EndianReader.cs:line 580
   at AssetRipper.Export.Modules.Shaders.ShaderBlob.ShaderSubProgram.Read(AssetReader reader) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.Modules.Shader\ShaderBlob\ShaderSubProgram.cs:line 84
   at AssetRipper.Export.Modules.Shaders.ShaderBlob.ShaderSubProgramBlob.ReadSegment(AssetReader reader, Int32 segment) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.Modules.Shader\ShaderBlob\ShaderSubProgramBlob.cs:line 58
   at AssetRipper.Export.Modules.Shaders.ShaderBlob.ShaderSubProgramBlob.ReadBlob(AssetCollection shaderCollection, Byte[] compressedBlob, UInt32 offset, UInt32 compressedLength, UInt32 decompressedLength, Int32 segment) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.Modules.Shader\ShaderBlob\ShaderSubProgramBlob.cs:line 47
   at AssetRipper.Export.Modules.Shaders.ShaderBlob.ShaderSubProgramBlob.Read(AssetCollection shaderCollection, Byte[] compressedBlob, AssetList`1 offsets, AssetList`1 compressedLengths, AssetList`1 decompressedLengths) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.Modules.Shader\ShaderBlob\ShaderSubProgramBlob.cs:line 31
   at AssetRipper.Export.Modules.Shaders.Extensions.ShaderExtensions.UnpackSubProgramBlobs(AssetCollection shaderCollection, AssetList`1 offsets, AssetList`1 compressedLengths, AssetList`1 decompressedLengths, Byte[] compressedBlob) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.Modules.Shader\Extensions\ShaderExtensions.cs:line 85
   at AssetRipper.Export.Modules.Shaders.Extensions.ShaderExtensions.ReadBlobs(IShader shader) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.Modules.Shader\Extensions\ShaderExtensions.cs:line 14
   at AssetRipper.Export.Modules.Shaders.IO.ShaderWriter..ctor(Stream stream, IShader shader, Func`2 exporterInstantiator) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.Modules.Shader\IO\ShaderWriter.cs:line 19
   at AssetRipper.Export.UnityProjects.Shaders.USCShaderExporter.ExportBinary(IShader shader, Stream stream, Func`2 exporterInstantiator) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.UnityProjects\Shaders\USCShaderExporter.cs:line 82
   at AssetRipper.Export.UnityProjects.Shaders.USCShaderExporter.Export(IExportContainer container, IUnityObjectBase asset, String path) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.UnityProjects\Shaders\USCShaderExporter.cs:line 45
   at AssetRipper.Export.UnityProjects.AssetExportCollection`1.ExportInner(IExportContainer container, String filePath, String dirPath) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.UnityProjects\AssetExportCollection.cs:line 69
   at AssetRipper.Export.UnityProjects.AssetExportCollection`1.Export(IExportContainer container, String projectDirectory) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.UnityProjects\AssetExportCollection.cs:line 28
   at AssetRipper.Export.UnityProjects.ProjectExporter.Export(GameBundle fileCollection, IEnumerable`1 files, CoreConfiguration options) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.UnityProjects\ProjectExporter.cs:line 277
   at AssetRipper.Export.UnityProjects.ProjectExporter.Export(GameBundle fileCollection, CoreConfiguration options) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.UnityProjects\ProjectExporter.cs:line 183
   at AssetRipper.Export.UnityProjects.Ripper.ExportProject(String exportPath, Func`2 filter, Action`1 onBeforeExport) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.UnityProjects\Ripper.cs:line 186
   at AssetRipper.GUI.Managers.UIExportManager.ExportInternal(Ripper ripper, MainWindowViewModel vm, String toRoot, IUnityObjectBase asset, Action onSuccess, Action`1 onError) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.GUI\Managers\UIExportManager.cs:line 61
```
2. using Dissassembly mode to export 1 shader, I also get the same crash during export:
```
System.IO.EndOfStreamException: Can't read 1869367076-byte string because there are only 393832 bytes left in the stream
   at AssetRipper.IO.Endian.EndianReader.ReadString(Int32 length) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.IO.Endian\EndianReader.cs:line 217
   at AssetRipper.Assets.IO.Reading.AssetReader.ReadString() in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Assets\IO\Reading\AssetReader.cs:line 22
   at AssetRipper.IO.Endian.EndianReader.ReadStringArray(Boolean allowAlignment) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.IO.Endian\EndianReader.cs:line 588
   at AssetRipper.IO.Endian.EndianReader.ReadStringArray() in F:\_DEV2\AssetRipper-master\Source\AssetRipper.IO.Endian\EndianReader.cs:line 580
   at AssetRipper.Export.Modules.Shaders.ShaderBlob.ShaderSubProgram.Read(AssetReader reader) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.Modules.Shader\ShaderBlob\ShaderSubProgram.cs:line 84
   at AssetRipper.Export.Modules.Shaders.ShaderBlob.ShaderSubProgramBlob.ReadSegment(AssetReader reader, Int32 segment) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.Modules.Shader\ShaderBlob\ShaderSubProgramBlob.cs:line 58
   at AssetRipper.Export.Modules.Shaders.ShaderBlob.ShaderSubProgramBlob.ReadBlob(AssetCollection shaderCollection, Byte[] compressedBlob, UInt32 offset, UInt32 compressedLength, UInt32 decompressedLength, Int32 segment) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.Modules.Shader\ShaderBlob\ShaderSubProgramBlob.cs:line 47
   at AssetRipper.Export.Modules.Shaders.ShaderBlob.ShaderSubProgramBlob.Read(AssetCollection shaderCollection, Byte[] compressedBlob, AssetList`1 offsets, AssetList`1 compressedLengths, AssetList`1 decompressedLengths) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.Modules.Shader\ShaderBlob\ShaderSubProgramBlob.cs:line 31
   at AssetRipper.Export.Modules.Shaders.Extensions.ShaderExtensions.UnpackSubProgramBlobs(AssetCollection shaderCollection, AssetList`1 offsets, AssetList`1 compressedLengths, AssetList`1 decompressedLengths, Byte[] compressedBlob) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.Modules.Shader\Extensions\ShaderExtensions.cs:line 85
   at AssetRipper.Export.Modules.Shaders.Extensions.ShaderExtensions.ReadBlobs(IShader shader) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.Modules.Shader\Extensions\ShaderExtensions.cs:line 14
   at AssetRipper.Export.Modules.Shaders.IO.ShaderWriter..ctor(Stream stream, IShader shader, Func`2 exporterInstantiator) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.Modules.Shader\IO\ShaderWriter.cs:line 19
   at AssetRipper.Export.UnityProjects.Shaders.ShaderDisassemblyExporter.ExportBinary(IShader shader, Stream stream, Func`2 exporterInstantiator) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.UnityProjects\Shaders\ShaderDisassemblyExporter.cs:line 61
   at AssetRipper.Export.UnityProjects.Shaders.ShaderDisassemblyExporter.Export(IExportContainer container, IUnityObjectBase asset, String path) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.UnityProjects\Shaders\ShaderDisassemblyExporter.cs:line 24
   at AssetRipper.Export.UnityProjects.AssetExportCollection`1.ExportInner(IExportContainer container, String filePath, String dirPath) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.UnityProjects\AssetExportCollection.cs:line 69
   at AssetRipper.Export.UnityProjects.AssetExportCollection`1.Export(IExportContainer container, String projectDirectory) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.UnityProjects\AssetExportCollection.cs:line 28
   at AssetRipper.Export.UnityProjects.ProjectExporter.Export(GameBundle fileCollection, IEnumerable`1 files, CoreConfiguration options) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.UnityProjects\ProjectExporter.cs:line 277
   at AssetRipper.Export.UnityProjects.ProjectExporter.Export(GameBundle fileCollection, CoreConfiguration options) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.UnityProjects\ProjectExporter.cs:line 183
   at AssetRipper.Export.UnityProjects.Ripper.ExportProject(String exportPath, Func`2 filter, Action`1 onBeforeExport) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.Export.UnityProjects\Ripper.cs:line 186
   at AssetRipper.GUI.Managers.UIExportManager.ExportInternal(Ripper ripper, MainWindowViewModel vm, String toRoot, IUnityObjectBase asset, Action onSuccess, Action`1 onError) in F:\_DEV2\AssetRipper-master\Source\AssetRipper.GUI\Managers\UIExportManager.cs:line 61
```
3. When I choose YAML asset mode:
   The YAML shader is exported fine, BUT, when I import it in Unity, it has errors:

```
Unable to parse file Assets/Shader TEST/Custom_PlayerColorShader.asset: [Parser Failure at line 610: Expect ':' between key and value within mapping]
```

Here are the relevant lines in the YAML shader:

```
          m_ParameterBlobIndices:
          - []
          - []
          - []
          - - 0
            - 1
            - 2 <- This is line 610 ... 
            - 2
```

[AssetRipper Decompile.log](https://github.com/AssetRipper/AssetRipper/files/12657581/AssetRipper.Decompile.log)
[AssetRipper Dissassembly.log](https://github.com/AssetRipper/AssetRipper/files/12657583/AssetRipper.Dissassembly.log)
[AssetRipper YAML.log](https://github.com/AssetRipper/AssetRipper/files/12657586/AssetRipper.YAML.log)
[Custom_PlayerColorShader.asset.txt](https://github.com/AssetRipper/AssetRipper/files/12657588/Custom_PlayerColorShader.asset.txt)
[PlayerColorShader OK 2020.3.35.asset.txt](https://github.com/AssetRipper/AssetRipper/files/12657685/PlayerColorShader.OK.2020.3.35.asset.txt)

It's the first time I encounter issues with YAML shaders, hence this report.

Using a previous version of the same game, compiled with Unity 2020.3.35, the YAML shaders are exported fine and unity is happy importing them (although I still have the same crashes using decompile or dissassembly modes).

-> So Something must have changed with the unity shader compilation between 2020.3.35 and 2021.3.30.......

Unrelated side note: I had a peek at the Electron UI: looks really promising !!! I like it very much ;)

Kind regards,
Z

**Comment by ds5678** [comment]
Hi Z,

For 1 and 2, you should comment those details on the existing issues for shader decompilation.

For 3, a file diff would be helpful.

About the Electron UI, I'm glad you like it. I'm about ready to drop Avalonia and do a prerelease. It still has a few small issues, but I want to show it off to the world. 😄

**Comment by Zbuffers** [comment]
Hmm not sure about how to generate a diff file, any hints ?

**Comment by ds5678** [comment]
I initially thought the double dash might be invalid yaml, but I checked with [YAMLlint](https://www.yamllint.com/), and it's fine.

I apologize about asking for the file diff. I had just woken up and missed that you had already attached the files I wanted to compare.

I don't know what's going on. It might be a bug on that unity version.

**Comment by Zbuffers** [comment]
Hi,

I got the same errors (can't export disassembly, and YAML shader erros in Unity) for another game (unity 2021.3.22),
So I did a little investigation on this. (invalid YAML Shader in unity 2021)

I think the problem is the shader serialization version (from 2 to 3).

Valid YAML shader header:
```
%YAML 1.1
%TAG !u! tag:unity3d.com,2011:
--- !u!48 &4800000
Shader:
  serializedVersion: 2 <- VERSION 2 IS OK (unity 2020)
  m_ObjectHideFlags: 0
  m_CorrespondingSourceObject: {fileID: 0}
  m_PrefabInstance: {fileID: 0}
  m_PrefabAsset: {fileID: 0}
  m_Name: Custom/PlayerColorShader
```
Invalid YAML shader header:
```
%YAML 1.1
%TAG !u! tag:unity3d.com,2011:
--- !u!48 &4800000
Shader:
  serializedVersion: 3 <- !!! <- VERSION 3 IS NOT OK (unity 2021)
  m_ObjectHideFlags: 0
  m_CorrespondingSourceObject: {fileID: 0}
  m_PrefabInstance: {fileID: 0}
  m_PrefabAsset: {fileID: 0}
  m_Name: Custom/PlayerColorShader
```

First, Unity reports a wrong line number for the error.
  In the example above: (from the attached file: Custom_PlayerColorShader.asset
   The error is on line 608, not 610 (stupid Unity...)
```
          m_ParameterBlobIndices:
          - []
          - []
          - []
          - - 0 <- line 608 ... This is where the error comes from, needs a value here something like (- - 0: VALUE)
            - 1
            - 2 <- line 610 ... wrong error line number reported by Unity
            - 2
```

Double dash is indeed a valid syntax (I think it is the first line of an array of arrays), BUT, it needs a value
here is a valid example (from line 424): 
`- - m_BlobIndex: 8`

Taking a look at the first 'progVertex' section (also applies to progFragment  sections), 
the 'old' (OK shader, serializedVersion: 2) uses 'm_SubPrograms' arrays
while the new (invalid shader, serializedVersion: 3), uses 'm_PlayerSubPrograms' arrays (and sets m_SubPrograms to empty ('[]')

m_SubPrograms and m_PlayerSubPrograms have a different layout and content...

So from what I see, it looks like the problem is the difference between serializedVersion 2 and 3, leading to 
the way progVertex and progFragment are serialized from m_SubPrograms to m_PlayerSubPrograms.

I have no idea as to how to fix this, I just hope that these infos can help you (or any contributors) improve shader exports.

Kind regards,
Z

**Comment by ds5678** [comment]
Serialized version 3 is correct for:
* 2021.3.12f1 and the remainder of 2021
* 2022.2.0b10 and the remainder of 2022
* 2023.1.0a11 and everything afterwards

Your line 608 error is most likely #695.

**Comment by Zbuffers** [comment]
Ha OK,
Do you think #695 could be the reason why shader dissassembly leads to crash during export ?

**Comment by ds5678** [comment]
No, the disassembly and decompilation errors are likely caused by either #524 or #726. #695 is completely unrelated.

**Comment by ds5678** [comment]
I would like to close this issue. All aspects of it appear to be covered by other issues.

**Comment by Zbuffers** [comment]
Sure ;)
Close it.

**Comment by GZHYBFHHJ** [comment]
It seems that Decompilation/Disassembly errors are caused by ShaderBlob structure changes in 2021.3

**Comment by GZHYBFHHJ** [comment]
According to my further analysis of 2021.3+ Shader YAML and blob data, I found:
In Shader YAML, `m_SubPrograms` in `prog*` has been replaced by `m_PlayerSubPrograms` and `m_ParameterBlobIndices`. Both of these point to the entry indices in the blob data.

In the blob data, ShaderSubProgram is divided into two parts: the part from `ProgramType` to `bindCount` (`PlayerSubProgram`), and `ParamGroups` (`ParameterBlob`), which all start with `BA 75 0A 0C`.
